spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,689 @@
+
+Parameters _AppHash_, _ConsensusParams_, and _ValidatorUpdates_ are used by Tendermint to compute various hashes in
+the block header that will finally be part of the proposal.
+
+* Property 2 [`PrepareProposal`, no-header-changes] When the blockchain is in next-block execution
+  mode, $p$'s Application cannot make any changes to the following parameters in `ResponsePrepareProposal`:
+  _AppHash_, _ConsensusParams_, _ValidatorUpdates_.
+
+In practical terms, Property 2 implies that Tendermint will ignore those parameters in `ResponsePrepareProposal`.
+
+* Property 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$,
+  if $q$'s Tendermint calls `RequestProcessProposal` on $v'_p$,
+  $q$'s Application returns Accept in `ResponseProcessProposal`.
+
+Property 3 makes sure that blocks proposed by correct processes _always_ pass the correct receiving process's
+`ProcessProposal` check.
+On the other hand, if there is a deterministic bug in `PrepareProposal` or `ProcessProposal` (or in both),
+strictly speaking, this makes all processes that hit the bug byzantine. This is a problem in practice,
+as very often validators are running the Application from the same codebase, so potentially _all_ would
+likely hit the bug at the same time. This would result in most (or all) processes prevoting `nil`, with the
+serious consequences on Tendermint's liveness that this entails.
+
+* Property 4 [`ProcessProposal`, determinism-1]: For any correct process $p$, and any arbitrary block $v'$,
+  if $p$'s Tendermint calls `RequestProcessProposal` on $v'$ at height $h$,
+  then $p$'s Application's acceptance or rejection exclusively depends on $v'$ and $s_{p,h-1}$.
+
+* Property 5 [`ProcessProposal`, determinism-2]: For any two correct processes $p$ and $q$, and any arbitrary block $v'$,
+  if $p$'s (resp. $q$'s) Tendermint calls `RequestProcessProposal` on $v'$ at height $h$,
+  then $p$'s Application accepts $v'$ if and only if $q$'s Application accepts $v'$.
+  Note that this Property follows from Property 4 and the Agreement property of consensus.
+
+Properties 4 and 5 ensure that all correct processes will react in the same way to a proposed block, even
+if the proposer is Byzantine. However, `ProcessProposal` may contain a bug that renders the
+acceptance or rejection of the block non-deterministic, and therefore prevents processes hitting
+the bug from fulfilling Properties 4 or 5 (effectively making those processes Byzantine).
+In such a scenario, Tendermint's liveness cannot be guaranteed.
+Again, this is a problem in practice if most validators are running the same software, as they are likely
+to hit the bug at the same point. There is currently no clear solution to help with this situation, so
+the Application designers/implementors must proceed very carefully with the logic/implementation
+of `ProcessProposal`. As a general rule `ProcessProposal` _should_ always accept the block.
```

Comment:
Also here we could add a note that this is a target for testing and verification.

Reply:
I guess this comment applies to all the properties ;-)

Reply:
Yes. And I also had an idea of a possible tool that would use these properties. Let's talk like